vsphere/vsphere.go

```diff
@@ -732,3 +742,134 @@ func isExponentialError(err error) bool {
 	}
 	return false
 }
+
+type progressLogger struct {
+	prefix string
+
+	wg sync.WaitGroup
+
+	sink chan chan progress.Report
+	done chan struct{}
+}
+
+func newProgressLogger(prefix string) *progressLogger {
+	p := &progressLogger{
+		prefix: prefix,
+
+		sink: make(chan chan progress.Report),
+		done: make(chan struct{}),
+	}
+
+	p.wg.Add(1)
+
+	go p.loopA()
+
+	return p
+}
+
+func createProgressLogger(prefix string) *progressLogger {
+	return newProgressLogger(prefix)
+}
+
+type OutputFlag struct {
+	common
+
+	JSON bool
+	XML  bool
+	TTY  bool
+	Dump bool
+	Out  io.Writer
+
+	formatError  bool
+	formatIndent bool
+}
+
+// Type to help flags out with only registering/processing once.
+type common struct {
+	register sync.Once
+	process  sync.Once
+}
+
+func (c *common) RegisterOnce(fn func()) {
+	c.register.Do(fn)
+}
+
+func (c *common) ProcessOnce(fn func() error) (err error) {
+	c.process.Do(func() {
+		err = fn()
+	})
+	return err
+}
+
+// loopA runs before Sink() has been called.
+func (p *progressLogger) loopA() {
+	var err error
+
+	defer p.wg.Done()
+
+	tick := time.NewTicker(100 * time.Millisecond)
+	defer tick.Stop()
+
+	called := false
+
+	for stop := false; !stop; {
+		select {
+		case ch := <-p.sink:
+			err = p.loopB(tick, ch)
+			stop = true
+			called = true
+		case <-p.done:
+			stop = true
+		case <-tick.C:
+			line := fmt.Sprintf("\r%s", p.prefix)
+			logrus.Infof(line)
+		}
+	}
+
+	if err != nil && err != io.EOF {
+		logrus.Error(fmt.Sprintf("\r%sError: %s\n", p.prefix, err))
+	} else if called {
+		logrus.Infof(fmt.Sprintf("\r%sOK\n", p.prefix))
+	}
+}
+
+// loopA runs after Sink() has been called.
+func (p *progressLogger) loopB(tick *time.Ticker, ch <-chan progress.Report) error {
+	var r progress.Report
+	var ok bool
+	var err error
+
+	for ok = true; ok; {
+		select {
+		case r, ok = <-ch:
+			if !ok {
+				break
+			}
+			err = r.Error()
```

Comment:
Don't we need to check r != nil here just like in the tick.C case ?
Also if we get an error shouldn't we exit out of this loop ?

Reply:
removed the progress feature all together.